Prompt the user to refresh the page if their session expires

kumar303 opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/2238

This is a baby step toward helping users cope with expired sessions. It's definitely not ideal but it's a quick patch that may be worth landing. 

<img width="361" alt="screenshot 2017-04-10 16 31 03" src="https://cloud.githubusercontent.com/assets/55398/24883921/b7ff3460-1e0c-11e7-85cd-dfc8a1039526.png">

Here are some things to do later:

* We can probably set a timer and anticipate when a session will expire and change the state. This is a harder patch and will be done in https://github.com/mozilla/addons-frontend/issues/2237 Even after doing that patch, it will still be possible to get a signature expired result though (because the timer won't be exact).
* ~~We can switch to detecting a code instead of the 'Signature expired' string, which is https://github.com/mozilla/addons-server/issues/5122~~
* ~~I thought it would be nice to show a better message like 'Your session has expired. You need to log in again' but this was difficult because we only have `i18n` objects inside components. Maybe I can think of a way.~~

tofumatt commented on src/core/reducers/errors.js:
```diff
@@ -12,7 +12,10 @@ import { gettext } from 'core/utils';
 function getMessagesFromError(error) {
-  let messages = [gettext('An unexpected error occurred')];
+  let data = {
+    messages: [gettext('An unexpected error occurred')],
```
Oh, so this isn't (and wasn't ever) translated. 😢

kumar303 replied:
Argh. We need to fix https://github.com/mozilla/addons-frontend/issues/2071

tofumatt commented on src/core/reducers/errors.js:
```diff
@@ -41,12 +44,23 @@ function getMessagesFromError(error) {
+      error.response.data.detail &&
+      // TODO: switch to checking for a code after
+      // https://github.com/mozilla/addons-server/issues/5122
+      error.response.data.detail === 'Signature has expired.'
```
This should at least be in a `gettext` call so we can grep for it, but I would really like for us to translate this message.

Also it should be changed. At least "Your session has expired." instead of "signature" which is confusing and less common terminology.

tofumatt replied:
Seems like, looking through the rest of the code, just an error code here that the component maps to a string, would be better and let us translate it.

tofumatt commented on src/core/reducers/errors.js:
```diff
@@ -41,12 +44,23 @@ function getMessagesFromError(error) {
+      error.response.data.detail === 'Signature has expired.'
+    ) {
+      // In this case, the only way to recover from the error is to
+      // refresh the page. It is impossible to fully avoid this state.
```
Is there any reason not to refresh the page for the user in this case? Forms are often saved by browsers but we should test that. Maybe it could be a post-merge new issue/patch.

kumar303 replied:
Reloading the page is pretty disruptive. It's hard to know for sure that the user won't lose any work so I didn't want to do it automatically. I think hitting this use case will be uncommon so I don't think we need to optimize it too much.

kumar303 commented:
@tofumatt thanks for the review -- could you take another look? I addressed everything and luckily https://github.com/mozilla/addons-server/issues/5122 has already landed so I switched to using codes.

kumar303 commented:
I updated the screenshot since the message is better now

tofumatt commented on src/core/reducers/errors.js:
```diff
@@ -12,14 +12,21 @@ import { gettext } from 'core/utils';
 function getMessagesFromError(error) {
-  let messages = [gettext('An unexpected error occurred')];
+  let data = {
+    code: undefined,
```
I feel like this should be its own code (`UNKNOWN_ERROR`) and that way the component can use real `i18n.gettext` for it.

I'm guessing this path should never happen but I think we might as well handle it correctly.

tofumatt commented on src/core/reducers/errors.js:
```diff
@@ -12,14 +12,21 @@ import { gettext } from 'core/utils';
  */
 function getMessagesFromError(error) {
-  let messages = [gettext('An unexpected error occurred')];
+  let data = {
```
`data` as a variable name is always a bit of a code smell to me. I know it's not a big difference but even `errorMessageData` or `errorData` would help it feel a bit less generic...

tofumatt commented on src/ui/components/ErrorList/index.js:
```diff
@@ -0,0 +1,67 @@
+    const items = [];
+
+    messages.forEach((msg) => {
+      let msgString = msg;
```
I think we can write out `message` here in both vars instead of `msg`.

tofumatt commented on src/core/constants.js:
```diff
@@ -174,3 +174,11 @@ export const GLOBAL_EVENT_STATUS_MAP = {
+// Generic error codes.
+export const ERROR_UNKNOWN = 'ERROR_UNKNOWN';
+// API error codes. These values match the error codes defined here:
+// http://addons-server.readthedocs.io/en/latest/topics/api/overview.html#unauthorized-and-permission-denied
```
Hyper nitpick but this can be HTTPS
